README.md
[![npm version](https://badge.fury.io/js/react-flip-page.svg)](https://badge.fury.io/js/react-flip-page)
[![Build Status](https://travis-ci.org/darenju/react-flip-page.svg?branch=master)](https://travis-ci.org/darenju/react-flip-page)
[![codecov](https://codecov.io/gh/darenju/react-flip-page/branch/master/graph/badge.svg)](https://codecov.io/gh/darenju/react-flip-page)
[![Greenkeeper badge](https://badges.greenkeeper.io/darenju/react-flip-page.svg)](https://greenkeeper.io/)
[![JavaScript Style Guide](https://img.shields.io/badge/code%20style-airbnb-brightgreen.svg)](https://github.com/airbnb/javascript)

# react-flip-page

> **DISCLAIMER**: This package is in no way related to nor endorsed by Flipboard, Inc. nor [flipboard.com](http://www.flipboard.com). This is just a showcase of HTML5 & CSS3 effect implemented with React.

This package allows you to use the cool Flipboard page swipe effect in your React.js apps.

> Please keep in mind that static content will work best with this component. Content like videos and related can produce undesired effects.

It has a `responsive` option, so you can possibly cover your entire page with it!

![Demo GIF](https://raw.githubusercontent.com/darenju/react-flip-page/master/demo.gif)

You can play with [this demo](http://darenju.me/react-flip-page/).

## Install

Installation is pretty straight-forward, as you just have to `npm install` this package:

```
npm install --save react-flip-page
```

Then, you can require the module with any way you like, let it be webpack or something else.

## Usage

This package consists of one single component that does all the work. Simply throw a `FlipPage` component with some children that will be the content.

```html
<FlipPage>
  <article>
    <h1>My awesome first article</h1>
    <p>My awesome first content</p>
  </article>
  <article>
    <h1>My wonderful second article</h1>
    <p>My wonderful second content</p>
  </article>
  <article>
    <h1>My excellent third article</h1>
    <p>My excellent third content</p>
  </article>
</FlipPage>
```

### Props

There are a few properties that define the behaviour of the component, here they are:

| Prop | Type | Default | Role |
|------|------|---------|------|
| `orientation` | `string` | `vertical` | Orientation of swipes. `vertical` or `horizontal` for respectively up/down swipes and left/right swipes |
| `uncutPages` | `boolean` | `false` | If `true`, the pages will be allowed to overflow through the container. The original effect is to keep everything inside the container, but you can set this to `true` to have a more "bookish" effect. |
| `animationDuration` | `number` | `200` | Duration in ms of the fold/unfold animation |
| `treshold` | `number` | `10` | Distance in px to swipe before the gesture is activated |
| `maxAngle` | `number` | `45` | Angle of the page when there's nothing to display before/after |
| `maskOpacity` | `number` | `0.4` | Opacity of the masks that covers the underneath content |
| `perspective` | `string` | `130em` | Perspective value of the page fold effect. The bigger, the less noticeable |
| `pageBackground` | `string` | `#fff` | Background of the pages. This can be overriden in individual pages by styling the component |
| `firstComponent` | `element` | `null` | Component that will be displayed under the first page |
| `lastComponent` | `element` | `null` | Component that will be displayed under the last page |
| `showHint` | `bool` | `false` | Indicates if the component must hint the user on how it works. Setting this to `true` will lift the bottom of the page 1s after the component is mounted, for 1s |
| `showSwipeHint` | `bool` | `false` | Indicates if the component must hint the user on how it works. Setting this to `true` will show an example of gesture to switch pages |
| `showTouchHint` | `bool` | `false` | Indicates if the component must hint the user on how it works. Setting this to `true` will show a pointer indicating where to click to switch pages. Works with
`flipOnTouch` |
| `style` | `object` | `{}` | Additional style for the flipboard |
| `height` | `number` | `480` | Height for the flipboard |
| `width` | `number` | `320` | Width for the flipboard |
| `onPageChange` | `function` |   | Callback when the page has been changed. Parameters: `pageIndex`, `direction` |
| `onStartPageChange` | `function` |   | Callback when the page starts to change. Parameters: `oldPageIndex`, `direction` |
| `onStartSwiping` | `function` |   | Callback when the user starts swiping |
| `onStopSwiping` | `function` |   | Callback when the user stops swiping |
| `className` | `string` | `''` | Optional CSS class to be applied on the container |
| `loopForever` | `boolean` | `false` | If `true` flipping after the last page will return to the first (and visa-versa) |
| `flipOnTouch` | `boolean` | `false` | If `true`, the user can flip pages by touching/clicking a top/bottom or left/right zone. These zones have CSS classes: `rfp-touchZone`, `rfp-touchZone-previous` and `rfp-touchZone-next` so that you can style them |
| `flipOnTouchZone` | `number` | `210` | Percentage of dimensions of the zone to touch/click to flip pages |
| `disableSwipe` | `boolean` | `false` | If `true`, users can't use the swipe feature to switch pages while `flipOnTouch` is enabled. Make sure you enable `flipOnTouch` so they can switch pages, or provide buttons binded to Methods |
| `responsive` | `boolean` | `false` | If `true`, the component will be responsive, meaning it will take all the available space. Place the component in a container before to make sure it is visible |
| `startAt` | `number` | `0` | Default start position of the component |
| `reverse` | `boolean` | `false` | If `true`, the user must swip in reverse order: he must swipe down/right to see the next page, and up/left to see the previous page. |
| `swipeImmune` | `array` | `[]` | This array holds the CSS class names that the user can not initiate a swipe gesture from. |
| `noShadow` | `boolean` | `false` | This disables the inset drop shadow on the inside of the flipping pages. |

## Methods

There are currently three methods that can be called on the component. To call them, you can use the
`ref` attribute in React:

```javascript
<FlipPage ref={(component) => { this.flipPage = component; }}>
  ...
</FlipPage>

this.flipPage.gotoPreviousPage();
```

### `gotoPreviousPage()`

This method triggers the effect and switches to the previous page, if possible.

### `gotoNextPage()`

This method triggers the effect and switches to the next page, if possible.

### `gotoPage(page)`

This methods positions the component to the wanted page index. The `page` argument should be between `0` and the number of pages. If not, a `RangeError` will be thrown. Also note that this does **not** call the `onPageChange` nor the `onStartPageChange` callback.

## Contribute

Since this is an open source project and it's far from perfect, contribution is welcome. Fork the repository and start working on your fix or new feature. Remember, it's good practice to work in your own branch, to avoid painful merge conflicts.

Once you think your work is ready, fire a [pull request](https://github.com/darenju/react-flip-page/pulls) with an understandable description of what you're bringing to the project. If it's alright, chances are high your work will be merged!

## Donate

This project takes some of my time, and I do it for free. If you're kind enough, you can show your support for my work by making a small donation here:

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](darenju@live.com)

I would very much appreciate it!


CHANGELOG.md

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 1.6.0

### Added

- Added a `noShadow` property that disables the inset drop shadow (which is in fact a gradient) on the inside of the flipping pages.

## 1.5.3

### Fixed

- Fixed the method to get the mouse/touch position.

## 1.5.2

### Fixed

- Fixed an event confusion in the `swipeImmune` feature.

## 1.5.1

### Fixed

- Fixed the `swipeImmune` detection condition.

## 1.5.0

### Added

- Added a `swipeImmune` property thanks to **@slevy85**. This property is an arary of CSS classes that makes the elements targeted by those classes immune to the swipe. This means the user can not initiate a swipe gesture from those elements.

### Fixed

- The `flipOnTouch` buttons were not working anymore due to some asynchronous refactoring.

## 1.4.0

### Added

- Added a `reverse` property thanks to **@slevy85**. The property allows the user to swipe pages in reverse order: the first page is visible, to see the second page, he has to swipe down instead of up.

### Fixed

- Tests were not working anymore. Added `babel-core@^7.0.0-bridge.0` and `babel-jest` in `devDependencies`.

## 1.3.0

### Fixed

- Fixed the swiping not being correctly registered.

## 1.2.2

### Fixed

- Fixed `gotoPreviousPage` & `gotoNextPage` method spam.
- Fixed right side of component being cut off.

## 1.2.1

### Fixed

- Condition in `gotoPage` method for checking inbounds of index was wrong, did not allow to go to first page (index `0`).

## 1.2.0

### Changed

- Call `onStartSwiping` after treshold is reached.

## 1.1.0

### Added

- Added `onStartPageChange` event. This event is triggered when the move gesture begins, in opposition to `onPageChange` event which is triggered after the gesture is complete.

## 1.0.0

### Added

- Added the list of contributors to the project, finalizing v1.0.0.

## 0.18.0

### Added

- Added a `gotoPage` method that takes an index of page in parameter. Places the component to the correct page.

## 0.17.0

### Changed

- Changed the way page components are rendered. To limit number of renders, the component now uses `PureComponent`s wrapper (higher-order components) for actual pages.

## 0.16.0

### Added

- New property `startAt`.

## 0.15.1

### Fixed

- Fixed `onPageChange` not passing the correct page number.

## 0.15.0

### Added

- Improve code coverage for `<FlipPage />` from `74%` to `99%`.
- `onStartSwiping` and `onEndSwiping` callback.

## 0.14.2

### Added

- Improve code coverage for `<FlipPage />` from `59%` to `74%`:
	- `mouseLeave()`: 100% covered
	- `reset()`: 100% covered
	- `props.orientation`: 100% covered
	- `moveGesture()`: 90% covered (WIP)

### Changed

- Change `eslintrc` run config, as it was given different error outputs based on the OS of the developer. It was basically change `eslint src/**/*.js* --quiet` to `eslint --ext .jsx --ext .js --quiet src/`.

## 0.14.1

### Fixed

- `button`s and `a` were not clickable.

## 0.14.0

### Added

- `responsive` property (default to `false`) to enable responsive mode.
- Improve code coverage for `<FlipPage />`:
	- `getHeight()`: 100% covered
	- `getHalfHeight()`: 100% covered
	- `getWidth()`: 100% covered
	- `getHalfWidth()`: 100% covered
	- `isLastPage()`: 100% covered
	- `isFirstPage()`: 100% covered
	- `incrementPage()`: 100% covered
	- `decrementPage()`: 100% covered
	- `doNotMove()`: 100% covered
	- `startMoving()`: 100% covered
- Fix some lint issues on `index.jsx`
- Add codecov to the project (@darenju still needs setup it to his github repository)

### Fixed

- Replace the code style badge url fron `standardjs` to `airbnb`.

## 0.13.1
### Added

- Added mask on previous/next parts. This way if you go to the drop area, you will see the shadow of the page. This only happened when lifting the page.

### Fixed

- Shadows were not displayed on some parts when swiping. The fix for non-clickable links had caused this.

## 0.13.0
### Added

- `showTouchHint` property that shows a pointer indicating where to click or touch to switch pages.

### Changed

- `flipOnTouchAllowDrag` has been renamed to `disableSwipe`. By default, the swipe is always allowed with `flipOnTouch`, unless `disableSwipe` is specified.
- `showTouchHint` has been renamed to `showSwipeHint` to avoid confusion.
- Changed default `flipOnTouchZone` from `20` to `10`.

### Removed

- Removed `touchHintTimeout`, `touchHintTimeoutTimer`, `showTouchHint`, `hintVisible` as they were useless.

## 0.12.2
### Added

- `flipOnTouchAllowDrag` propery allowing the user to keep the dragging feature even though `flipOnTouch`
is active.

## 0.12.1
### Added

- Class names on touch zones.

## 0.12.0
### Added

- Add `flipOnTouch` feature. Zones to touch/click are defined by the `flipOnTouchZone` property.

## 0.11.8
### Added

- Add `.gitattributes` for windows developers
- Add Jest setup
- Add Enzyme setup
- Add `code coverage` report file
- Add `check-dependencies` to `prebuild` and `pretest` lifecycle. This lib checks if the developer has the up-to-date dependencies inside its node_modules directory, as described in the `package.json` file.
- Add TravisCI and Greenkeeper initial setup (@darenju still needs to setup each of these in his repository)
- Add `pre-commit` which runs lint before each commit

### Removed

- Remove generated code from VCS, as a good practice. The generated code should not be versioned, as it can be checked at [Best Practices for Node.js Development](https://devcenter.heroku.com/articles/node-best-practices#only-git-the-important-bits)

### Fixed


lib/flippage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var m = function m() {
  for (var _len = arguments.length, objs = Array(_len), _key = 0; _key < _len; _key++) {
    objs[_key] = arguments[_key];
  }

  return Object.assign.apply(Object, [{}].concat(objs));
};

var FlipPage = function (_Component) {
  _inherits(FlipPage, _Component);

  function FlipPage(props) {
    _classCallCheck(this, FlipPage);

    var _this = _possibleConstructorReturn(this, (FlipPage.__proto__ || Object.getPrototypeOf(FlipPage)).call(this, props));

    _this.state = {
      page: 0, // current index of page
      startY: -1, // start position of swipe
      diff: 0, // difference between last swipe position and current position
      timestamp: 0, // time elapsed between two swipes
      angle: 0, // rotate angle of half page
      rotate: 0, // absolute value of above, limited to 45° if necessary
      direction: '', // original swipe direction
      lastDirection: '', // last registered swipe direction
      bottomStyle: {}, // transform style of bottom half
      topStyle: {} // transform style of top half
    };

    // binding events
    _this.startMoving = _this.startMoving.bind(_this);
    _this.moveGesture = _this.moveGesture.bind(_this);
    _this.stopMoving = _this.stopMoving.bind(_this);
    _this.reset = _this.reset.bind(_this);

    _this.transition = 'transform ' + _this.props.animationDuration / 1000 + 's ease-in-out';
    return _this;
  }

  _createClass(FlipPage, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (this.props.showHint) {
        setTimeout(function () {
          return _this2.showHint();
        }, 1000);
      }
    }
  }, {
    key: 'isLastPage',
    value: function isLastPage() {
      return this.state.page + 1 === _react.Children.count(this.props.children);
    }
  }, {
    key: 'isFirstPage',
    value: function isFirstPage() {
      return this.state.page === 0;
    }
  }, {
    key: 'getHeight',
    value: function getHeight() {
      return this.props.height + 'px';
    }
  }, {
    key: 'getHalfHeight',
    value: function getHalfHeight() {
      return this.props.height / 2 + 'px';
    }
  }, {
    key: 'getWidth',
    value: function getWidth() {
      return this.props.width + 'px';
    }
  }, {
    key: 'showHint',
    value: function showHint() {
      var _this3 = this;

      this.setState({
        bottomStyle: {
          transition: this.transition
        }
      }, function () {
        _this3.setState({
          bottomStyle: {
            transition: _this3.transition,
            transform: 'perspective(' + _this3.props.perspective + ') rotateX(30deg)'
          }
        });

        setTimeout(function () {
          _this3.setState({
            bottomStyle: {
              transition: _this3.transition
            }
          });
        }, 1000);
      });
    }
  }, {
    key: 'startMoving',
    value: function startMoving(e) {
      e.preventDefault();

      var posY = e.pageY || e.touches[0].pageY;

      this.setState({ startY: posY });
    }
  }, {
    key: 'moveGesture',
    value: function moveGesture(e) {
      e.preventDefault();

      var posY = e.pageY || e.touches[0].pageY;

      if (this.state.startY !== -1) {
        var diff = posY - this.state.startY;
        var angle = diff / 250 * 180;
        var useMaxAngle = false;
        if (this.state.direction === 'up') {
          useMaxAngle = this.isLastPage();
        } else if (this.state.direction === 'down') {
          useMaxAngle = this.isFirstPage();
        }
        var rotate = Math.min(Math.abs(angle), useMaxAngle ? this.props.maxAngle : 180);

        // determine direction to prevent two-directions swipe
        if (this.state.direction === '' && Math.abs(diff) > this.props.treshold) {
          var direction = '';

          if (diff < 0) {
            direction = 'up';
          } else if (diff > 0) {
            direction = 'down';
          }

          this.setState({ direction: direction });
        }

        // set the last direction
        var lastDirection = this.state.lastDirection;
        if (this.state.diff > diff) {
          lastDirection = 'up';
        } else if (this.state.diff < diff) {
          lastDirection = 'down';
        }

        this.setState({
          angle: angle,
          rotate: rotate,
          timestamp: Date.now(),
          diff: diff,
          lastDirection: lastDirection
        });

        // flip bottom
        if (diff < 0 && this.state.direction === 'up') {
          this.setState({ angle: angle, bottomStyle: {
              transform: 'perspective(' + this.props.perspective + ') rotateX(' + rotate + 'deg)'
            } });
        } else if (diff > 0 && this.state.direction === 'down') {
          this.setState({ angle: angle, topStyle: {
              transform: 'perspective(' + this.props.perspective + ') rotateX(-' + rotate + 'deg)',
              zIndex: 2 // apply a z-index to pop over the back face
            } });
        }
      }
    }
  }, {
    key: 'stopMoving',
    value: function stopMoving(e) {
      var _this4 = this;

      var delay = Date.now() - this.state.timestamp;

      var goNext = !this.isLastPage() && (this.state.angle <= -90 || delay <= 20 && this.state.direction === 'up' && this.state.lastDirection === 'up');
      var goPrevious = !this.isFirstPage() && (this.state.angle >= 90 || delay <= 20 && this.state.direction === 'down' && this.state.lastDirection === 'down');

      // reset everything
      this.reset();
      this.setState({
        bottomStyle: {
          transition: this.transition,
          transform: goNext ? 'perspective(' + this.props.perspective + ') rotateX(180deg)' : ''
        },

        topStyle: {
          transition: this.transition,
          transform: goPrevious ? 'perspective(' + this.props.perspective + ') rotateX(-180deg)' : '',
          zIndex: goPrevious ? 2 : 'auto'
        }
      }, function () {
        // load the next item if it was requested
        if (goNext) {
          setTimeout(function () {
            _this4.setState({
              bottomStyle: {},
              page: _this4.state.page + 1
            });
          }, _this4.props.animationDuration);
        } else if (goPrevious) {
          // or load the previous item
          setTimeout(function () {
            _this4.setState({
              topStyle: {},
              page: _this4.state.page - 1
            });
          }, _this4.props.animationDuration);
        }
      });
    }
  }, {
    key: 'reset',
    value: function reset() {
      this.setState({
        startY: -1,
        angle: 0,
        rotate: 0,
        direction: '',
        lastDirection: '',
        bottomStyle: {
          transition: this.transition
        },
        topStyle: {
          transition: this.transition
        }
      });
    }
  }, {
    key: 'renderPage',
    value: function renderPage(page, key) {
      var height = this.getHeight();
      var halfHeight = this.getHalfHeight();
      var width = this.getWidth();
      var gradientTop = '0 -100px 100px -100px rgba(0,0,0,0.25) inset';
      var gradientBottom = '0 100px 100px -100px rgba(0,0,0,0.25) inset';

      var pageItem = (0, _react.cloneElement)(page, {
        style: Object.assign({}, page.props.style, {
          height: height
        })
      });

      var style = {
        container: {
          display: this.state.page === key ? 'block' : 'none',
          height: height,
          overflow: 'hidden',
          position: 'relative',
          width: width
        },
        part: {
          height: halfHeight,
          left: 0,
          position: 'absolute',
          width: width
        },
        visiblePart: {
          transformStyle: 'preserve-3d'
        },
        top: {
          top: 0,
          transformOrigin: 'bottom center'
        },
        bottom: {
          bottom: 0,
          transformOrigin: 'top center'
        },
        face: {
          backfaceVisibility: 'hidden',
          WebkitBackfaceVisibility: 'hidden',
          height: halfHeight,
          left: 0,
          position: 'absolute',
          top: 0,
          transformStyle: 'preserve-3d',
          width: width
        },
        back: {
          transform: 'rotateX(180deg)'
        },
        before: {
          top: 0
        },
        after: {
          bottom: 0
        },
        cut: {
          background: this.props.pageBackground,
          height: halfHeight,
          overflow: 'hidden',
          position: 'relative',
          width: width
        },
        pull: {
          marginTop: '-' + halfHeight
        },
        gradient: {
          position: 'absolute',
          left: 0,
          right: 0,
          bottom: 0,
          top: 0,
          transition: 'box-shadow ' + this.props.animationDuration / 1000 + 's ease-in-out'
        },
        gradientBottomFace: {
          boxShadow: this.state.direction === 'up' ? gradientBottom : ''
        },
        gradientTopFace: {
          boxShadow: this.state.direction === 'down' ? gradientTop : ''
        },
        gradientBottomBack: {
          boxShadow: this.state.direction === 'up' ? gradientTop : ''
        },
        gradientTopBack: {
          boxShadow: this.state.direction === 'down' ? gradientBottom : ''
        },
        mask: {
          position: 'absolute',
          top: 0,
          left: 0,
          right: 0,
          bottom: 0,
          backgroundColor: '#000',
          opacity: this.state.direction !== '' ? Math.max(this.props.maskOpacity - Math.abs(this.state.rotate) / 90 * this.props.maskOpacity, 0) : 0
        }
      };

      var beforeItem = !this.isFirstPage() ? this.props.children[this.state.page - 1] : this.props.firstComponent;

      var afterItem = !this.isLastPage() ? this.props.children[this.state.page + 1] : this.props.lastComponent;

      var container = style.container,
          part = style.part,
          visiblePart = style.visiblePart,
          top = style.top,
          bottom = style.bottom,
          face = style.face,
          back = style.back,
          before = style.before,
          after = style.after,
          cut = style.cut,
          pull = style.pull,
          gradient = style.gradient,
          gradientBottomBack = style.gradientBottomBack,
          gradientTopBack = style.gradientTopBack,
          gradientBottomFace = style.gradientBottomFace,
          gradientTopFace = style.gradientTopFace,
          mask = style.mask;


      return React.createElement(
        'div',
        {
          key: key,
          onMouseDown: this.startMoving,
          onTouchStart: this.startMoving,
          onMouseMove: this.moveGesture,
          onTouchMove: this.moveGesture,
          onMouseUp: this.stopMoving,
          onTouchEnd: this.stopMoving,
          onMouseLeave: this.reset,
          style: container
        },
        React.createElement(
          'div',
          { style: m(part, before, cut) },
          beforeItem,
          React.createElement('div', { style: mask })
        ),
        React.createElement(
          'div',
          { style: m(part, after, cut) },
          React.createElement(
            'div',
            { style: pull },
            afterItem
          ),
          React.createElement('div', { style: mask })
        ),
        React.createElement(
          'div',
          { style: m(part, visiblePart, top, this.state.topStyle) },
          React.createElement(
            'div',
            { style: face },
            React.createElement(
              'div',
              { style: cut },
              pageItem
            ),
            React.createElement('div', { style: m(gradient, gradientTopFace) })
          ),
          React.createElement(
            'div',
            { style: m(face, back) },
            React.createElement(
              'div',
              { style: cut },
              React.createElement(
                'div',
                { style: pull },
                beforeItem
              )
            ),
            React.createElement('div', { style: m(gradient, gradientTopBack) })
          )
        ),
        React.createElement(
          'div',
          { style: m(part, visiblePart, bottom, this.state.bottomStyle) },
          React.createElement(
            'div',
            { style: face },
            React.createElement(
              'div',
              { style: cut },
              React.createElement(
                'div',
                { style: pull },
                pageItem
              )
            ),
            React.createElement('div', { style: m(gradient, gradientBottomFace) })
          ),
          React.createElement(
            'div',
            { style: m(face, back) },
            React.createElement(
              'div',
              { style: m(part, after, cut) },
              afterItem
            ),
            React.createElement('div', { style: m(gradient, gradientBottomBack) })
          )
        )
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _this5 = this;

      var style = m(this.props.style, {
        height: this.getHeight(),
        position: 'relative',
        width: this.getWidth()
      });

      // all the pages are rendered once, to prevent glitching
      // (React would reload the child page and cause a image glitch)
      return React.createElement(
        'div',
        { style: style },
        _react.Children.map(this.props.children, function (page, key) {
          return _this5.renderPage(page, key);
        })
      );
    }
  }]);

  return FlipPage;
}(_react.Component);

Flipboard.defaultProps = {
  animationDuration: 200,
  treshold: 10,
  maxAngle: 45,
  maskOpacity: 0.4,
  perspective: '130em',
  pageBackground: '#fff',
  firstComponent: null,
  lastComponent: null,
  showHint: false,
  style: {}
};

Flipboard.propTypes = {
  animationDuration: _propTypes2.default.number,
  treshold: _propTypes2.default.number,
  maxAngle: _propTypes2.default.number,
  maskOpacity: _propTypes2.default.number,
  perspective: _propTypes2.default.string,
  pageBackground: _propTypes2.default.string,
  firstComponent: _propTypes2.default.element,
  lastComponent: _propTypes2.default.element,
  showHint: _propTypes2.default.bool,
  style: _propTypes2.default.object
};

exports.default = FlipPage;
